runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkStatefulDoFnFunction.java

```diff
@@ -227,12 +224,14 @@ public void open(Configuration parameters) {
     // deserialization method. However, this is a hack, and we want to properly initialize the
     // options where they are needed.
     FileSystems.setDefaultPipelineOptions(serializedOptions.get());
+    metricContainer = new FlinkMetricContainer(getRuntimeContext());
     doFnInvoker = DoFnInvokers.tryInvokeSetupFor(dofn);
   }
 
   @Override
   public void close() throws Exception {
     try {
+      metricContainer.registerMetricsForPipelineResult();
```

Comment:
Is it guaranteed that the accumulators created in `close` will be captured in the terminated Flink job? Is it covered by existing tests?

Reply:
Yes, I've checked the Flink code. Have a look at https://github.com/apache/flink/blob/16b635790c77c522a42bac2dbc98011b28f10096/flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java#L1461 When the task is removed, this will send a final accumulator snapshot. The metrics tests which rely on the MetricsPusher would also fail if that did not work correctly.